src/Storages/ObjectStorage/StorageObjectStorage.cpp

```diff
@@ -0,0 +1,476 @@
+    }
+
+private:
+    ObjectStoragePtr object_storage;
+    ConfigurationPtr configuration;
+    std::shared_ptr<StorageObjectStorageSource::IIterator> iterator_wrapper;
+
+    const ReadFromFormatInfo info;
+    const NamesAndTypesList virtual_columns;
+    const std::optional<DB::FormatSettings> format_settings;
+    const String name;
+    const bool need_only_count;
+    const size_t max_block_size;
+    const size_t num_streams;
+    const bool distributed_processing;
+
+    void createIterator(const ActionsDAG::Node * predicate)
+    {
+        if (iterator_wrapper)
+            return;
+        auto context = getContext();
+        iterator_wrapper = StorageObjectStorageSource::createFileIterator(
+            configuration, object_storage, distributed_processing,
+            context, predicate, virtual_columns, nullptr, context->getFileProgressCallback());
+    }
+};
+}
+
+void StorageObjectStorage::read(
+    QueryPlan & query_plan,
+    const Names & column_names,
+    const StorageSnapshotPtr & storage_snapshot,
+    SelectQueryInfo & query_info,
+    ContextPtr local_context,
+    QueryProcessingStage::Enum /*processed_stage*/,
+    size_t max_block_size,
+    size_t num_streams)
+{
+    updateConfiguration(local_context);
+    if (partition_by && configuration->withPartitionWildcard())
+    {
+        throw Exception(ErrorCodes::NOT_IMPLEMENTED,
+                        "Reading from a partitioned {} storage is not implemented yet",
+                        getName());
+    }
+
+    const auto read_from_format_info = prepareReadingFromFormat(
+        column_names, storage_snapshot, supportsSubsetOfColumns(local_context));
+    const bool need_only_count = (query_info.optimize_trivial_count || read_from_format_info.requested_columns.empty())
+        && local_context->getSettingsRef().optimize_count_from_files;
+
+    auto read_step = std::make_unique<ReadFromObjectStorageStep>(
+        object_storage,
+        configuration,
+        getName(),
+        column_names,
+        getVirtualsList(),
+        query_info,
+        storage_snapshot,
+        format_settings,
+        distributed_processing,
+        read_from_format_info,
+        need_only_count,
+        local_context,
+        max_block_size,
+        num_streams);
+
+    query_plan.addStep(std::move(read_step));
+}
+
+SinkToStoragePtr StorageObjectStorage::write(
+    const ASTPtr & query,
+    const StorageMetadataPtr & metadata_snapshot,
+    ContextPtr local_context,
+    bool /* async_insert */)
+{
+    updateConfiguration(local_context);
+    const auto sample_block = metadata_snapshot->getSampleBlock();
+    const auto & settings = configuration->getQuerySettings(local_context);
+
+    if (configuration->withGlobsIgnorePartitionWildcard())
+    {
+        throw Exception(ErrorCodes::DATABASE_ACCESS_DENIED,
+                        "Path '{}' contains globs, so the table is in readonly mode",
+                        configuration->getPath());
+    }
```

Comment:
Let's probably also disallow writing when path is archive. Because right now such write will lead to unexpected things:
```
insert into function s3('http://localhost:11111/test/archive.zip::data.jsonl') select 42 as x settings s3_truncate_on_insert=1
```

It will owerwrite file `archive.zip` and will write data in JSONEachRow format in this file

Reply:
And the same for truncate I guess